File: packages/escrow/src/scores.ts
/**
 * Reading a league's `Scores` account back, and checking it against the rules
 * members signed.
 *
 * ## Why this exists
 *
 * The account carries four things the derivation needs and the score poster must
 * not be free to choose — the tiebreaker chain, the playoff week window, the
 * first-round bye count, and whether third place is played. Pick the tiebreakers
 * and you pick the best-regular-season-record prize holder, which is the same
 * defect as posting a standing outright.
 *
 * **The program cannot check them.** A rules hash is 32 opaque bytes to it, so
 * it has no way to know whether the terms in the account are the terms in the
 * document. That is exactly the position `initialize_league` is in, and the
 * answer is the same one: compare off-chain, and gate something on the result.
 *
 * What it is gated on is **the draw**. `drawDraftOrder` already refuses until
 * the chain says the season started; it now also refuses a `Scores` account that
 * disagrees with the signed rules. A league that never draws has no order, no
 * roster, no schedule and nothing to score — so this is a gate rather than a
 * courtesy, the same way `joinLeague` refusing an unanchored league is what
 * makes `anchorTermMismatches` load-bearing rather than advisory.
 *
 * ## The oracle key is compared too, since schemaVersion 8
 *
 * It was not, for as long as it took to move the schema. Until then the
 * commissioner who created the account chose freely who could post their
 * league's scores, which is the whole of the attack `docs/SETTLEMENT.md` §6
 * describes: name your own key, post the scores that make you champion, and
 * every other check in the system passes.
 */

import type { Program } from "@coral-xyz/anchor";
import type { PublicKey } from "@solana/web3.js";

import { scoresPda } from "./program.js";
import type { RostrEscrow } from "./types.js";

/**
 * The wire value of each tiebreaker, mirroring `Tiebreaker` in `derive.rs`.
 *
 * **Written out rather than derived from the union's declaration order**, which
 * is the lesson `PRIZE_ORDER` paid for: the two agree today and nothing makes
 * them, so a reordered union would silently renumber the chain and change who is
 * seed 1. There is a test pinning these.
 */
export const TIEBREAKER_DISCRIMINANTS: Readonly<Record<string, number>> = {
  WIN_PCT: 0,
  POINTS_FOR: 1,
  HEAD_TO_HEAD: 2,
  POINTS_AGAINST: 3,
  LOWEST_TEAM_ID: 4,
};

export interface OnChainRosterEntry {
  /** Lower-case hex without dashes. Compare with `uuidToHex`. */
  readonly teamIdHex: string;
  /** Base58. */
  readonly wallet: string;
}

export interface OnChainScores {
  readonly address: PublicKey;
  readonly league: string;
  readonly oracle: string;
  readonly roster: readonly OnChainRosterEntry[];
  readonly tiebreakers: readonly number[];
  readonly playoffWeeks: readonly number[];
  readonly regularSeasonWeeks: number;
  readonly playoffTeams: number;
  readonly firstRoundByes: number;
  readonly thirdPlace: boolean;
  /** Set once the prizes have been paid. */
  readonly settled: boolean;
  /** Bit `w` set means week `w + 1` is frozen. */
  readonly finalizedWeeks: number;
  /**
   * Unix seconds, as a decimal string, or `"0"` while nothing is finalised.
   *
   * A string for the same reason `refundUnlockAt` is one: it is an `i64`, and
   * `BN.toNumber()` throws above 2^53 rather than letting a comparison run.
   */
  readonly lastFinalizedAt: string;
}

/** A Postgres uuid as the 32 lower-case hex characters the program stores. */
export function uuidToHex(uuid: string): string {
  return uuid.replace(/-/g, "").toLowerCase();
}

/**
 * Read a league's `Scores` account, or `null` when it does not exist.
 *
 * Missing is the ordinary case for every league before its commissioner creates
 * one, so it is a return value rather than a throw.
 */
export async function fetchOnChainScores(
  program: Program<RostrEscrow>,
  leagueId: string,
): Promise<OnChainScores | null> {
  const address = scoresPda(leagueId);
  const account = await program.account["scores"]?.fetchNullable(address);
  if (!account) return null;

  const raw = account as {
    league: PublicKey;
    oracle: PublicKey;
    roster: { teamId: number[]; wallet: PublicKey }[];
    tiebreakers: ArrayLike<number>;
    playoffWeeks: ArrayLike<number>;
    regularSeasonWeeks: number;
    playoffTeams: number;
    firstRoundByes: number;
    thirdPlace: boolean;
    settled: boolean;
    finalizedWeeks: number;
    lastFinalizedAt: { toString(): string };
  };

  return {
    address,
    league: raw.league.toBase58(),
    oracle: raw.oracle.toBase58(),
    roster: raw.roster.map((entry) => ({
      teamIdHex: Array.from(entry.teamId)
        .map((byte) => byte.toString(16).padStart(2, "0"))
        .join(""),
      wallet: entry.wallet.toBase58(),
    })),
    tiebreakers: Array.from(raw.tiebreakers),
    playoffWeeks: Array.from(raw.playoffWeeks),
    regularSeasonWeeks: raw.regularSeasonWeeks,
    playoffTeams: raw.playoffTeams,
    firstRoundByes: raw.firstRoundByes,
    thirdPlace: raw.thirdPlace,
    settled: raw.settled,
    finalizedWeeks: raw.finalizedWeeks,
    lastFinalizedAt: raw.lastFinalizedAt.toString(),
  };
}

/** The subset of a rule set this comparison reads. */
export interface ScoreTermRules {
  /** `null` for a free league, which has no settlement and no oracle. */
  readonly pot: { readonly settlementOracle: string } | null;
  readonly schedule: {
    readonly regularSeasonWeeks: number;
    readonly playoffWeeks: readonly number[];
    readonly playoffTeams: number;
    readonly byeSeeds: number;
    readonly tiebreakers: readonly string[];
  };
}

export interface ExpectedScoreTerms {
  /** Base58, or `null` for a free league. */
  readonly oracle: string | null;
  readonly tiebreakers: readonly number[];
  readonly playoffWeeks: readonly number[];
  readonly regularSeasonWeeks: number;
  readonly playoffTeams: number;
  readonly firstRoundByes: number;
  readonly thirdPlace: boolean;
  readonly teamCount: number;
}

/**
 * What a league's `Scores` account must say, given its signed rules and the size
 * of the field that actually formed.
 *
 * **`firstRoundByes` depends on both**, which is why `teamCount` is a parameter
 * rather than something read out of the rules. The frozen `byeSeeds` applies
 * only when the playoff field is the size it was frozen for; a smaller league
 * takes whatever byes its own size needs to form a valid round. `playoffState`
 * in `@rostr/db` makes the same choice for the same reason, and the two must not
 * disagree — a wrong bye count changes who plays whom in a bracket that decides
 * the pot.
 */
export function expectedScoreTerms(
  rules: ScoreTermRules,
  teamCount: number,
): ExpectedScoreTerms {
  const field = Math.min(rules.schedule.playoffTeams, teamCount);

  return {
    oracle: rules.pot?.settlementOracle ?? null,
    tiebreakers: rules.schedule.tiebreakers.map((name) => {
      const value = TIEBREAKER_DISCRIMINANTS[name];
      // Refused rather than defaulted: a tiebreaker nobody recognises must not
      // quietly become "win percentage", which is what index 0 would do.
      if (value === undefined) throw new Error(`unknown tiebreaker: ${name}`);
      return value;
    }),
    playoffWeeks: [...rules.schedule.playoffWeeks],
    regularSeasonWeeks: rules.schedule.regularSeasonWeeks,
    // The signed seat count, not the field that formed. `settle` caps it by the
    // roster itself, so a league that never filled still brackets correctly —
    // but the *stored* value has to be the one members agreed to, or a caller
    // could shrink the bracket by writing a smaller number.
    playoffTeams: rules.schedule.playoffTeams,
    firstRoundByes:
      field === rules.schedule.playoffTeams ? rules.schedule.byeSeeds : byesFor(field),
    // Always played, paid or not — `playsThirdPlace` in `@rostr/db`. A constant
    // rather than a rule field today; if it ever becomes one, this follows it.
    thirdPlace: true,
    teamCount,
  };
}

/** Byes needed to bring a field up to a power of two. Mirrors `byesFor`. */
function byesFor(field: number): number {
  let power = 1;
  while (power < field) power *= 2;
  return power - field;
}

/**
 * Everything the on-chain account says that the signed rules do not.
 *
 * Empty means it agrees. **Every mismatch is reported rather than the first**,
 * because they are frozen together and the account can never be rewritten — so
 * whoever has to recreate the league needs the whole list, not one item per
 * attempt. Same shape and same reasoning as `anchorTermMismatches`.
 */
export function scoresTermMismatches(
  onChain: OnChainScores,
  expected: ExpectedScoreTerms,
): string[] {
  const out: string[] = [];
  const same = (a: readonly number[], b: readonly number[]): boolean =>
    a.length === b.length && a.every((value, index) => value === b[index]);

  /*
    The oracle first, because it is the one that decides everything else.

    Whoever holds this key posts the scores the champion is derived from, so a
    key nobody signed is a league whose result was decided by a stranger — and
    every other check here would pass while that was true. It is compared
    against the signed document rather than against our own configuration for
    the same reason `anchorTermMismatches` compares the fee recipient that way:
    what matters is that members agreed to it, not that we recognise it.
  */
  if (expected.oracle !== null && onChain.oracle !== expected.oracle) {
    out.push(`settlementOracle: chain has ${onChain.oracle}, rules say ${expected.oracle}`);
  }

  if (!same(onChain.tiebreakers, expected.tiebreakers)) {
    out.push(
      `tiebreakers: chain has [${onChain.tiebreakers.join(", ")}], ` +
        `rules say [${expected.tiebreakers.join(", ")}]`,
    );
  }
  if (!same(onChain.playoffWeeks, expected.playoffWeeks)) {
    out.push(
      `playoffWeeks: chain has [${onChain.playoffWeeks.join(", ")}], ` +
        `rules say [${expected.playoffWeeks.join(", ")}]`,
    );
  }
  if (onChain.playoffTeams !== expected.playoffTeams) {
    out.push(
      `playoffTeams: chain has ${onChain.playoffTeams}, rules say ${expected.playoffTeams}`,
    );
  }
  if (onChain.regularSeasonWeeks !== expected.regularSeasonWeeks) {
    out.push(
      `regularSeasonWeeks: chain has ${onChain.regularSeasonWeeks}, ` +
        `rules say ${expected.regularSeasonWeeks}`,
    );
  }
  if (onChain.firstRoundByes !== expected.firstRoundByes) {
    out.push(
      `firstRoundByes: chain has ${onChain.firstRoundByes}, ` +
        `rules say ${expected.firstRoundByes}`,
    );
  }
  if (onChain.thirdPlace !== expected.thirdPlace) {
    out.push(`thirdPlace: chain has ${onChain.thirdPlace}, rules say ${expected.thirdPlace}`);
  }
  // A roster short of the field cannot pay everyone; one longer names somebody
  // who is not in the league. Both are unrecoverable once the season starts and
  // both are cheap to catch here.
  if (onChain.roster.length !== expected.teamCount) {
    out.push(
      `roster: chain has ${onChain.roster.length} teams, the league has ${expected.teamCount}`,
    );
  }

  return out;
}

File: programs/rostr-escrow/src/scores.rs
//! Scores, and who they belong to. G7.
//!
//! `derive.rs` and `bracket.rs` can work out who won given a roster and a list
//! of games. This is where those two things live, and it is the only account in
//! the program that anything writes to after creation.
//!
//! `docs/SETTLEMENT.md` is the design. The three ideas that decide the shape:
//!
//! ## 1. The roster is the payee list, and it is fixed before the season
//!
//! Nothing on-chain otherwise connects a team to a wallet — `Membership` knows a
//! wallet and a stake, and a team is a Postgres fact. Somebody has to attest the
//! pairing, and the only real question is **when**, because an attestation made
//! at payout time cannot be checked by anyone before the money is gone.
//!
//! So it is made here, before a single game is posted, and it is write-once.
//! Every member can check their own row for the whole season; if it is wrong,
//! nobody calls `start_season` and every stake is refundable 48 hours after the
//! draft.
//!
//! **And the wallet is derived, never supplied.** The caller passes one
//! `Membership` account per team and the wallet is read out of it, so a caller
//! with no way to *name* a wallet has no way to name somebody else's. That also
//! makes the funded check structural rather than a rule: a team's payee is a
//! wallet that demonstrably staked in this league.
//!
//! ## 2. Weeks are editable until finalised, and frozen after
//!
//! Write-once per week is the obvious design and it is a one-way ratchet: the
//! program holds no schedule (deliberately — see `initialize_league`), so it
//! cannot tell week 17's real result from week 17 posted in September, and a
//! single wrong index makes `compute_records` refuse **forever** with no
//! overwrite, no `close`, and after the upgrade burn no fix. A typo would cost
//! the pot.
//!
//! `finalize_week` is the explicit lock instead. It says the same thing
//! `RULES.md` §7 already says off-chain — a finalised week is never rescored —
//! without making a mistake terminal.
//!
//! ## 3. `finalized_at` is what makes the settlement hold possible
//!
//! Payout is illegal until seven days after the last week it needs was
//! finalised. That is the only bound in this design on a settlement oracle that
//! *works and lies*; everything else bounds one that is absent. See
//! `SETTLEMENT_HOLD_SECONDS`.

use anchor_lang::prelude::*;

use crate::derive::{MAX_TEAMS, MAX_WEEKS};
use crate::{EscrowError, League, Membership};

/// Most games one week can hold: everybody paired, nobody idle.
pub const MAX_GAMES_PER_WEEK: usize = MAX_TEAMS / 2;

/// Most links a tiebreaker chain may have. Five exist in `Tiebreaker`; the slack
/// is for appending one without a layout change.
pub const MAX_TIEBREAKERS: usize = 8;

/// Most weeks a bracket may span.
pub const MAX_PLAYOFF_WEEKS: usize = 8;

/// How long after the last needed week finalises before a payout may run.
///
/// **Seven days, and it is the only thing here that bounds a dishonest oracle.**
/// Losing the key, or never posting, is already bounded: the timelock returns
/// every stake, so members are delayed rather than robbed. None of that helps
/// when the key works and posts scores that are wrong on purpose, because that
/// happens long before the timelock opens.
///
/// Deriving the champion rather than being told it converts "post a winner" into
/// "post the scores that produce that winner" — the same authority, with a
/// public receipt. This is the gap between the receipt and the money, so that
/// the two-provider comparison `RULES.md` §7 promises can happen *before*
/// settlement rather than after it.
///
/// **It is not a veto.** Nobody gains the power to stop a payout; the hold
/// expires on its own. What it buys is time to notice, and the remedy for
/// noticing is to not send the payout at all and let the timelock refund
/// everyone — which needs no instruction and no authority, and is why it is the
/// remedy chosen.
///
/// Seven days spends about a sixth of the window between "settlement is legal"
/// and "refunds open", and matches the stat-correction number members already
/// know from finalisation.
pub const SETTLEMENT_HOLD_SECONDS: i64 = 7 * 24 * 60 * 60;

/// One team's place in the payee list.
#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, Default, InitSpace)]
pub struct RosterEntry {
    /// The team's Postgres UUID, raw bytes. Feeds `LOWEST_TEAM_ID`, the last
    /// link in the default tiebreaker chain — so this is not merely a label, it
    /// participates in the seeding and cannot be replaced by the wallet.
    pub team_id: [u8; 16],
    /// Where this team's prize is paid. Read from a funded `Membership` at
    /// creation, never supplied by the caller.
    pub wallet: Pubkey,
}

/// One completed game.
///
/// **A bye is not representable, deliberately.** `compute_records` treats a bye
/// as "not a game, not a point, not a record", so posting one would change
/// nothing; making it unrepresentable removes a way to write a row that means
/// nothing and can still be got wrong.
#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, Default, InitSpace)]
pub struct PostedGame {
    pub home: u8,
    pub away: u8,
    pub home_milli_points: u32,
    pub away_milli_points: u32,
}

#[account]
#[derive(InitSpace)]
pub struct Scores {
    pub league: Pubkey,
    pub bump: u8,

    /// The only key that may post. **Compared against a `Signer`, never inferred
    /// from the instruction stack** — a program that identifies its caller
    /// through the `instruction_sysvar` breaks under a multisig, because that
    /// sysvar sees only top-level instructions and would return the multisig
    /// program's id. A stored key against a signer is unforgeable however deeply
    /// the call is nested, which is what lets this be a Squads vault address.
    pub oracle: Pubkey,

    #[max_len(MAX_TEAMS)]
    pub roster: Vec<RosterEntry>,

    // ---------------------------------------------------------------------
    // The frozen inputs the derivation needs and the oracle must not choose.
    //
    // They are here rather than on `League` because `League`'s layout is fixed
    // and this account did not exist yet. What binds them to what members signed
    // is not the program — it cannot read a rules hash — but the draw, which
    // refuses to draw an order for a league whose `Scores` terms disagree with
    // the signed document. A league that never draws never plays.
    //
    // If they were instruction arguments instead, whoever posts scores would
    // also pick the tiebreaker chain, and therefore the best-record prize
    // holder. That is the same defect as posting a standing.
    // ---------------------------------------------------------------------
    #[max_len(MAX_TIEBREAKERS)]
    pub tiebreakers: Vec<u8>,
    /// The weeks the playoff bracket may use, ascending.
    #[max_len(MAX_PLAYOFF_WEEKS)]
    pub playoff_weeks: Vec<u8>,
    /// Weeks 1..=this are the regular season, and only those feed the seeding.
    pub regular_season_weeks: u8,
    pub first_round_byes: u8,
    pub third_place: bool,

    /// Games by week. Index `w` holds week `w + 1`.
    #[max_len(MAX_WEEKS, MAX_GAMES_PER_WEEK)]
    pub games: Vec<Vec<PostedGame>>,

    /// Bit `w` set means week `w + 1` has been finalised and can never change.
    pub finalized_weeks: u32,

    /// When the most recent week was finalised. Zero until one is.
    ///
    /// Payout reads this, not the individual weeks: the hold runs from the
    /// *last* finalisation, so a late correction to any week restarts the clock
    /// that lets somebody notice it.
    pub last_finalized_at: i64,
}

impl Scores {
    fn week_index(&self, week: u8) -> Result<usize> {
        require!(
            week >= 1 && (week as usize) <= MAX_WEEKS,
            EscrowError::UnknownWeek
        );
        Ok(week as usize - 1)
    }

    /// Whether a week is frozen.
    pub fn is_finalized(&self, week: u8) -> bool {
        match self.week_index(week) {
            Ok(index) => self.finalized_weeks & (1u32 << index) != 0,
            Err(_) => false,
        }
    }

    /// The instant a payout may first run, or `None` while nothing is finalised.
    pub fn settlement_opens_at(&self) -> Option<i64> {
        if self.last_finalized_at == 0 {
            None
        } else {
            Some(self.last_finalized_at + SETTLEMENT_HOLD_SECONDS)
        }
    }
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone)]
pub struct InitializeScoresArgs {
    /// One per team, in any order — the derivation indexes by position, and the
    /// roster's own order is what those indices mean from here on.
    pub team_ids: Vec<[u8; 16]>,
    pub oracle: Pubkey,
    pub tiebreakers: Vec<u8>,
    pub playoff_weeks: Vec<u8>,
    pub regular_season_weeks: u8,
    pub first_round_byes: u8,
    pub third_place: bool,
}

#[derive(Accounts)]
pub struct InitializeScores<'info> {
    #[account(
        seeds = [b"league", league.league_id.as_ref()],
        bump = league.bump,
        // The commissioner creates this, not the oracle — the oracle key is
        // *content* of this account, so at creation there is nothing stored to
        // check a signer against. Anchoring it to the commissioner is what stops
        // a stranger front-running creation and writing a hostile roster, which
        // would deny settlement permanently since nothing closes this account.
        has_one = commissioner @ EscrowError::NotCommissioner,
    )]
    pub league: Box<Account<'info, League>>,

    #[account(
        init,
        payer = commissioner,
        space = 8 + Scores::INIT_SPACE,
        seeds = [b"scores", league.key().as_ref()],
        bump,
    )]
    pub scores: Box<Account<'info, Scores>>,

    #[account(mut)]
    pub commissioner: Signer<'info>,

    pub system_program: Program<'info, System>,
    // remaining_accounts: one `Membership` per entry in `team_ids`, same order.
}

#[derive(Accounts)]
pub struct PostWeek<'info> {
    #[account(
        mut,
        seeds = [b"scores", scores.league.as_ref()],
        bump = scores.bump,
        has_one = oracle @ EscrowError::NotTheOracle,
    )]
    pub scores: Box<Account<'info, Scores>>,
    pub oracle: Signer<'info>,
}

/// Write a payee roster and the terms the derivation runs under. Once.
///
/// Refuses a league with no pot: a free league has no vault, nothing to settle,
/// and no `Membership` to derive a wallet from.
pub fn initialize_scores(ctx: Context<InitializeScores>, args: InitializeScoresArgs) -> Result<()> {
    let league = &ctx.accounts.league;
    require!(league.has_pot, EscrowError::LeagueHasNoPot);
    // Before the season is declared started, because that is the moment the
    // roster stops being checkable and the failed-league refund shuts. Members
    // get the whole window between this and `start_season` to find their own
    // row, and `start_season` is what they withhold if it is wrong.
    require!(!league.started, EscrowError::AlreadyStarted);

    let count = args.team_ids.len();
    require!(count >= 2, EscrowError::FieldTooSmall);
    require!(count <= MAX_TEAMS, EscrowError::TooManyTeams);
    require!(
        args.tiebreakers.len() >= 1 && args.tiebreakers.len() <= MAX_TIEBREAKERS,
        EscrowError::NoTiebreakers
    );
    require!(
        !args.playoff_weeks.is_empty() && args.playoff_weeks.len() <= MAX_PLAYOFF_WEEKS,
        EscrowError::NotEnoughWeeks
    );
    require!(
        args.regular_season_weeks as usize <= MAX_WEEKS,
        EscrowError::UnknownWeek
    );
    require!(
        (args.first_round_byes as usize) < count,
        EscrowError::FieldTooSmall
    );
    /*
      And the oracle is not the commissioner.

      The commissioner writes this account, so without this they could name
      themselves and post the scores that decide a pot they hold a stake in —
      `docs/RULES.md` §9's "not permitted, ever: anything that touches a roster,
      a score, a standing, or the pot", in one transaction.

      Off-chain the key comes from server configuration and the draw compares it
      against the signed rules, which is the stronger check. This is the one that
      binds a caller who never touched our service, and it costs a comparison.
    */
    require_keys_neq!(
        args.oracle,
        league.commissioner,
        EscrowError::OracleIsCommissioner
    );
    require_keys_neq!(args.oracle, Pubkey::default(), EscrowError::OracleMissing);

    // Every discriminant has to be one this build understands, checked now
    // rather than in December: an unknown tiebreaker written here would make the
    // derivation refuse forever, on an account that can never be rewritten.
    for value in &args.tiebreakers {
        crate::derive::Tiebreaker::from_u8(*value)?;
    }

    // One `Membership` per team, in the same order. The wallet is read out of
    // it rather than taken as an argument — see the module docs.
    require!(
        ctx.remaining_accounts.len() == count,
        EscrowError::RosterIncomplete
    );

    let mut roster: Vec<RosterEntry> = Vec::with_capacity(count);
    for (index, team_id) in args.team_ids.iter().enumerate() {
        let info = &ctx.remaining_accounts[index];
        // Deserialised by hand rather than through `Account::try_from` — the
        // borrow would have to outlive this loop iteration, and `try_deserialize`
        // checks the discriminator anyway, which is the part that matters.
        require_keys_eq!(*info.owner, crate::ID, EscrowError::NotAMembership);
        let membership = {
            let data = info.try_borrow_data()?;
            Membership::try_deserialize(&mut &data[..])
                .map_err(|_| error!(EscrowError::NotAMembership))?
        };

        require_keys_eq!(
            membership.league,
            league.key(),
            EscrowError::MembershipLeagueMismatch
        );
        // Staked, and not since withdrawn. `deposited > 0` alone reads a
        // refunded member as funded — the four `Membership` states are set out
        // in CLAUDE.md and this is the one that catches people out.
        require!(membership.deposited > 0, EscrowError::NothingDeposited);
        require!(!membership.refunded, EscrowError::AlreadyRefunded);

        // The account really is this league's membership for this wallet, and
        // not some other account that happens to deserialize.
        let (expected, _) = Pubkey::find_program_address(
            &[
                b"membership",
                league.key().as_ref(),
                membership.member.as_ref(),
            ],
            ctx.program_id,
        );
        require_keys_eq!(info.key(), expected, EscrowError::MembershipMemberMismatch);

        // No wallet twice: two teams sharing a payee would send both prizes to
        // one person and leave a real member unpayable.
        require!(
            !roster
                .iter()
                .any(|e: &RosterEntry| e.wallet == membership.member),
            EscrowError::DuplicateRosterEntry
        );
        require!(
            !roster.iter().any(|e: &RosterEntry| e.team_id == *team_id),
            EscrowError::DuplicateRosterEntry
        );

        roster.push(RosterEntry {
            team_id: *team_id,
            wallet: membership.member,
        });
    }

    let scores = &mut ctx.accounts.scores;
    scores.league = league.key();
    scores.bump = ctx.bumps.scores;
    scores.oracle = args.oracle;
    scores.roster = roster;
    scores.tiebreakers = args.tiebreakers;
    scores.playoff_weeks = args.playoff_weeks;
    scores.regular_season_weeks = args.regular_season_weeks;
    scores.first_round_byes = args.first_round_byes;
    scores.third_place = args.third_place;
    scores.games = vec![Vec::new(); MAX_WEEKS];
    scores.finalized_weeks = 0;
    scores.last_finalized_at = 0;

    Ok(())
}

/// Replace a week's games. Legal until that week is finalised.
pub fn post_week(ctx: Context<PostWeek>, week: u8, games: Vec<PostedGame>) -> Result<()> {
    let scores = &mut ctx.accounts.scores;
    let index = scores.week_index(week)?;
    require!(!scores.is_finalized(week), EscrowError::WeekAlreadyFinal);
    require!(games.len() <= MAX_GAMES_PER_WEEK, EscrowError::TooManyGames);

    let teams = scores.roster.len() as u8;
    for game in &games {
        require!(game.home < teams, EscrowError::UnknownTeam);
        require!(game.away < teams, EscrowError::UnknownTeam);
        // A team cannot play itself, and the derivation would silently credit
        // both halves of the result to one record.
        require!(game.home != game.away, EscrowError::MalformedSchedule);
    }

    scores.games[index] = games;
    Ok(())
}

/// Freeze a week. Nothing can change it afterwards, and the settlement hold
/// restarts from now.
pub fn finalize_week(ctx: Context<PostWeek>, week: u8) -> Result<()> {
    let now = Clock::get()?.unix_timestamp;
    let scores = &mut ctx.accounts.scores;
    let index = scores.week_index(week)?;
    require!(!scores.is_finalized(week), EscrowError::WeekAlreadyFinal);

    scores.finalized_weeks |= 1u32 << index;
    // The hold runs from the most recent finalisation rather than from week 17
    // specifically, so a week corrected and frozen late restarts the window in
    // which somebody can notice it.
    scores.last_finalized_at = now;
    Ok(())
}
